### DISCUSSION TITLE
Different dual variable using the MOSEK and the Yalmip

### DISCUSSION THREAD

**Discussion by 0xF0F0F0F** [yalmip/YALMIP#932] [Problems using YALMIP] [answered]
Hi, I am trying to find the dual variable of 3rd inequality constraint of following problem:

y=sdpvar(N,1);     options = sdpsettings('solver','mosek');
obj = y.'*H*y + f.'*y;
const=[A*y>=cl; A*y<=cu; gamma>=norm(chol(R)*y)];
optimize(const,obj,options);
lambda_yalmip = dual(const(3));

which could provide fairly accurate dual variable: lambda. However, for a large scale problem, using the YALMIP is not fast enough. Therefore, I try to use MOSEK directly:

[r, res] = mosekopt('symbcon');      param.MSK_DPAR_INTPNT_CO_TOL_PFEAS = 1.0e-12;
prob.c      = [f', 1];  

prob.a = [sparse(A),sparse(size(A,1),1)];
prob.buc = cu;                 prob.blc = cl;
prob.blx = -inf(N+1,1);    prob.bux = inf(N+1,1);

prob.f  = [sparse(1,N+1); chol(R), sparse(N,1)];
prob.g = [gamma; zeros(N,1)];
prob.cones = [ res.symbcon.MSK_CT_QUAD N+1 ];                  

prob.f  = [prob.f ; [sparse(1,N), 1; sparse(1,N+1); chol(H), sparse(N,1)] ];
prob.g = [prob.g; [ 0; 0.5; zeros(N,1) ] ];
prob.cones = [prob.cones, res.symbcon.MSK_CT_RQUAD N+2 ];         

[~, res] = mosekopt('minimize echo(10)', prob, param);
lambda_mosek = res.sol.itr.doty(1); 

It is much faster, but I got less accurate result of dual variable: lambda. I am wondering, it is possible to make two "lambda"s identical? Many Thanks!

**Comment by johanlofberg** [comment, marked as answer]
Sounds weird that it would be (significantly slower) via YALMIP for such a simple model, so a reproducible example would be nice. It could be that you've switch the primal/side side of the problem interpretation compared to yalmip leading to a smaller but numerically more challenging problem or something like that

**Reply by johanlofberg** [reply]
and don't use the form y'*H*y for large-scale model and/or dense H, as that will lead to massive symbolic overhead. Hence change objective to e'*e with added constraint e == chol(H)*y, or use a norm representation which is to replace it with t and the socp norm([2*chol(H);1-t])<=1+t

**Reply by 0xF0F0F0F** [reply]
Thank you very much, Johan! It is now much faster!
